lib/matplotlib/axes/_axes.py

```diff
@@ -1971,8 +2000,48 @@ def bar(self, left, height, width=0.8, bottom=None, **kwargs):
         See also
         --------
         barh: Plot a horizontal bar plot.
+
         """
         kwargs = cbook.normalize_kwargs(kwargs, mpatches._patch_alias_map)
+
+        matchers = [
+            (lambda x, height, width, bottom, **kwargs:
+             (False, x, height, width, bottom, kwargs)),
+            (lambda x, height, width, **kwargs:
+             (False, x, height, width, None, kwargs)),
+            (lambda x, height, bottom, **kwargs:
+             (False, x, height, 0.8, bottom, kwargs)),
+            (lambda x, height, **kwargs:
+             (False, x, height, 0.8, None, kwargs)),
+
+            (lambda left, height, width, bottom, **kwargs:
+             (True, left, height, width, bottom, kwargs)),
+            (lambda left, height, width, **kwargs:
+             (True, left, height, width, None, kwargs)),
+            (lambda left, height, bottom, **kwargs:
+             (True, left, height, 0.8, bottom, kwargs)),
+            (lambda left, height, **kwargs:
+             (True, left, height, 0.8, None, kwargs)),
+        ]
+        for matcher in matchers:
+            try:
+                dp, x, height, width, y, kwargs = matcher(*args, **kwargs)
+            except TypeError:
+                # This can only come from a no-match as there is
+                # no other logic in the matchers.
+                continue
+            else:
+                break
+        else:
+            raise TypeError("Missing two required positional arguments: 'x'"
```

Comment:
Can you save the first exception and just reraise it here?  this will make the message correct in more cases (e.g. the user passes height as a kwarg but not x).

Reply:
The first one won't be right, it is the 4th one we want, but if you put the 4th one first it will hit and then we would have to handle the kwargs for the two optional ones seperatly (which might be better).

Reply:
nm